File: 2-D DP/String DP/Interleaving String/solution.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  bool isInterleave(string s1, string s2, string s3) {
    int m = s1.length(), n = s2.length(), k = s3.length();
    if (m + n != k)
      return false;

    vector<bool> dp(n + 1, false);
    dp[0] = true;

    for (int j = 1; j <= n; ++j) {
      dp[j] = dp[j - 1] && (s2[j - 1] == s3[j - 1]);
    }

    for (int i = 1; i <= m; ++i) {
      dp[0] = dp[0] && (s1[i - 1] == s3[i - 1]);
      for (int j = 1; j <= n; ++j) {
        bool fromTop = dp[j] && (s1[i - 1] == s3[i + j - 1]);
        bool fromLeft = dp[j - 1] && (s2[j - 1] == s3[i + j - 1]);
        dp[j] = fromTop || fromLeft;
      }
    }

    return dp[n];
  }
};

int main() {
  Solution sol;
  cout << "Test Case 1: "
       << (sol.isInterleave("aabcc", "dbbca", "aadbbcbcac") ? "True" : "False")
       << endl; // Expect True
  cout << "Test Case 2: "
       << (sol.isInterleave("aabcc", "dbbca", "aadbbbaccc") ? "True" : "False")
       << endl; // Expect False
  return 0;
}

File: 2-D DP/String DP/Interleaving String/solution.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False
            
        m, n = len(s1), len(s2)
        dp = [False] * (n + 1)
        dp[0] = True
        
        for j in range(1, n + 1):
            dp[j] = dp[j-1] and s2[j-1] == s3[j-1]
            
        for i in range(1, m + 1):
            dp[0] = dp[0] and s1[i-1] == s3[i-1]
            for j in range(1, n + 1):
                from_top = dp[j] and s1[i-1] == s3[i+j-1]
                from_left = dp[j-1] and s2[j-1] == s3[i+j-1]
                dp[j] = from_top or from_left
                
        return dp[n]

if __name__ == "__main__":
    sol = Solution()
    print(f"Test Case 1: {sol.isInterleave('aabcc', 'dbbca', 'aadbbcbcac')}") # True
    print(f"Test Case 2: {sol.isInterleave('aabcc', 'dbbca', 'aadbbbaccc')}") # False
